docs/proposals/20230706-yurtappconfigurationreplacement.md

```diff
@@ -0,0 +1,291 @@
+	metav1.TypeMeta `json:",inline"`
+	// Standard object's metadata
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	// Describe the object this replacement belongs
+	Subject Subject `json:"subject"`
+	// Describe detailed multi-region configuration of the subject above
+	Replacements []Replacement `json:"replacements"`
+}
+```
+### Architecture
+The whole architecture is shown below. 
+<img src = "../img/yurtappconfigurationreplacement/Architecture.png" width="800">
+
+### Implementation Details
+#### Deployment Mutating Webhook
+##### Prerequisites for webhook (Resolving circular dependency)
+Since YurtManager is deployed as a Deployment, the Deployment webhook and YurtManager create a circular dependency. 
+
+Solutions:
+1. Change YurtManager deploying method, like static pod
+2. Controller is responsible for both creating and updating. However, there will be a period of unavailability(wrong configuration information)
+3. Webhook's failurePolicy set to ignore(difficult to detect in the case of malfunction)
+4. YurtManager is in charge of managing the webhook, we can modify the internal implementation of YurtManager(Recommended)
+##### Workflow of mutating webhook
+1. If the intercepted Deployment's ownerReferences field is empty, filter it directly
+2. Find the corresponding YurtAppConfigurationReplacement resource by ownerReferences, if not, filter directly
+3. Find the replacements involved, get the corresponding configuration, and inject them into workloads. 
+
+Attention Points:
+1. Note that injection is implemented by recalculating the final configuration according to the YurtAppSet workload template and the watching YurtAppConfigurationReplacement
+2. The latter configuration always relpace the former. So the last configuration will really work
+#### YurtAppConfigurationReplacement Validating Webhook
+1. Verify that only one field of item is selected
+2. Verify that replicas and upgradeStrategy are selected only once in a replacement
+#### YurtAppConfigurationReplacement Controller
+##### Task 1
+1. Get update events by watching the YurtAppConfigurationReplacement resource
+2. Trigger the Deployment mutating webhook by modifying an annotation or label
+##### Task 2
+1. Get delete events(delete members of pools) by watching the YurtAppConfigurationReplacement resource
+2. Render the configuration according to the YurtAppSet workload template and the watching YurtAppConfigurationReplacement
+#### Advanced features
+Here is a scenario. Ten nodepools only need to be configured differently for configMap, it is not user-friendly to write ten replacements by using current scheme. 
+We add a new feature like wildcard to simplify the usage. We can name all the configMap as configMap-{{nodepool}} so that we can inject the configMap into nodepool in the suffix. 
+```yaml
+replacements:
+- pools:
+    *
+  items:
+  - configMap:
+      configMapSource: configMap-demo
+      configMapTarget: configMapName-{{nodepool}}
```

Comment:
Here, the configMapSource is the name of ConfigMap in the original template, the configMapTarget is newly written externally and will be injected into. So I think you got it backwards. 
Do you think "source" should correspond to the newly written configMap?

Reply:
I got it. I got them backwards. Thank you for your explanation.

Reply:
This may confirm that it is easy to look backwards, and then we can discuss it together.

Reply:
Ok, we can communicate it on Wednesday.

Reply:
Does the mode of Source and resource only support the **replace** rendering way and do not support injecting new configmap or PVC? Do we not consider supporting the **add** rendering way？

Reply:
It is true that the current scheme does not support the injection of new configMap. The current API may not be applicable. 
On the one hand, users can add key-value pairs to the configMapTarget. On the other hand, if one deployment need injection all of PVC, secret, and configMap, would it be more appropriate to create a separate new deployment rather than deploying in a YurtAppSet&YurtAppConfigurationReplacement way?
This is our @rambohe-ch understanding.